### ISSUE TITLE
Wrong diff when storing old data

### ISSUE THREAD

**Issue by bgervan** [Opus10/django-pghistory#171]
Hi, first of all, great package.

My use case:
I am storing the old data, to make it easier to revert to something. I am storing transaction_id in context, so I can revert all related changes in a request easily. (At least this is the goal)

If I am using the following settings:

```python
PGHISTORY_DEFAULT_TRACKERS = (
    pghistory.InsertEvent(),
    pghistory.UpdateEvent(row=pghistory.Old),
    pghistory.DeleteEvent(row=pghistory.Old),
)
```

the pgh_diff is wrong, as it fetch the current and previous data of the event. I have changed the raw sql to reach my goal to the this:

```sql
WITH pghistory_events AS (SELECT CONCAT('sample.ExampleModelEvent', ':', _pgh_obj_event.pgh_id) AS pgh_slug,
                                 _pgh_obj_event.pgh_id,
                                 _pgh_obj_event.pgh_created_at,
                                 _pgh_obj_event.pgh_label,
                                 _pgh_obj_event.pgh_obj_id,
                                 'sample.ExampleModelEvent'                                     AS pgh_model,
                                 'sample.ExampleModel'                                          AS pgh_obj_model,
                                 (SELECT JSONB_OBJECT_AGG(filtered.key, filtered.value)
                                  FROM (SELECT key,
                                               value
                                        FROM JSONB_EACH(_pgh_obj_event._curr_data::JSONB)) filtered
                                  WHERE filtered.key NOT LIKE 'pgh_%')                          AS pgh_data,
                                 (SELECT JSONB_OBJECT_AGG(filtered.key, filtered.value)
                                  FROM (SELECT key,
                                               value
                                        FROM JSONB_EACH(_pgh_obj_event._prev_data::JSONB)) filtered
                                  WHERE filtered.key NOT LIKE 'pgh_%')                          AS pgh_old_data,
                                 (SELECT JSONB_OBJECT_AGG(curr.key, array [prev.value, curr.value])
                                  FROM (SELECT key,
                                               value
                                        FROM JSONB_EACH(_pgh_obj_event._curr_data::JSONB)) curr
                                           LEFT OUTER JOIN (SELECT key,
                                                                   value
                                                            FROM JSONB_EACH(_pgh_obj_event._prev_data::JSONB)) prev
                                                           ON curr.key = prev.key
                                  WHERE curr.key NOT LIKE 'pgh_%'
                                    AND curr.value != prev.value
                                    AND prev IS NOT NULL)                                       AS pgh_diff,
                                 (SELECT JSONB_OBJECT_AGG(curr.key, array [prev.value, curr.value])
                                  FROM (SELECT key,
                                               value
                                        FROM JSONB_EACH(_pgh_obj_event._real_data::JSONB)) curr
                                           LEFT OUTER JOIN (SELECT key,
                                                                   value
                                                            FROM JSONB_EACH(_pgh_obj_event._prev_data::JSONB)) prev
                                                           ON curr.key = prev.key
                                  WHERE curr.key NOT LIKE 'pgh_%'
                                    AND curr.value != prev.value
                                    AND prev IS NOT NULL)                                       AS pgh_diff_to_real,
                                 _pgh_obj_event.pgh_context_id,
                                 _pgh_obj_event.pgh_context
                          FROM (SELECT pgh_id,
                                       pgh_created_at,
                                       pgh_label,
                                       row_to_json(_event)               AS _prev_data,
                                       COALESCE(
                                                       LEAD(row_to_json(_event))
                                                       OVER (PARTITION BY _event.pgh_obj_id ORDER BY _event.pgh_id),
                                                       (SELECT row_to_json(em)
                                                        FROM sample_examplemodel em
                                                        WHERE em.id = _event.pgh_obj_id)
                                       )                                 AS _curr_data,
                                       (SELECT row_to_json(em)
                                        FROM sample_examplemodel em
                                        WHERE em.id = _event.pgh_obj_id) AS _real_data,
                                       pgh_context_id,
                                       pgh_context,
                                       pgh_obj_id::TEXT
                                FROM sample_examplemodelevent _event
                                ORDER BY _event.pgh_id) _pgh_obj_event)
SELECT "pghistory_events"."pgh_slug",
       "pghistory_events"."pgh_model",
       "pghistory_events"."pgh_id",
       "pghistory_events"."pgh_created_at",
       "pghistory_events"."pgh_label",
       "pghistory_events"."pgh_data",
       "pghistory_events"."pgh_old_data",
       "pghistory_events"."pgh_diff",
       "pghistory_events"."pgh_diff_to_real",
       "pghistory_events"."pgh_context_id",
       "pghistory_events"."pgh_context",
       "pghistory_events"."pgh_obj_model",
       "pghistory_events"."pgh_obj_id"
FROM "pghistory_events"
WHERE "pghistory_events"."pgh_slug" = 'sample.ExampleModelEvent:17'
LIMIT 21
```

Important part is:
```sql
row_to_json(_event)  AS _prev_data,
COALESCE(
    LEAD(row_to_json(_event))
         OVER (PARTITION BY _event.pgh_obj_id ORDER BY _event.pgh_id),
        (SELECT row_to_json(em) FROM sample_examplemodel em WHERE em.id = _event.pgh_obj_id)
    )  AS _curr_data,
(SELECT row_to_json(em) FROM sample_examplemodel em WHERE em.id = _event.pgh_obj_id) AS _real_data,
```

Also added pgh_old_data, so it is easier to debug. In this setup:
- pgh_old_data is the current event data, with Old, this is the previous object state
- pgh_data is the next or the real data
- phg_diff is the diff between the current and next event data or the real object, depends on existence -> this is the real diff that should be implemented with pghistory.Old
- pgh_diff_to_real is the diff between the current state and the change prior to the event
- sample_examplemodel is the tracked model
- sample_examplemodelevent is the event model

Can we add changes like this optionally? As I understand, all these affect only the admin, and relevant only when using the pghistory.Old

The **phg_diff** is important, but the **diff_to_real** is also nice to have, as with that, we can show the diffs between the revert and current state before the user click on the revert. 

_EDIT:_ Added important sql sections separately for easier reading

**Comment by bgervan** [comment]
I have move forward a bit with the actual implementation:

```python
    def _get_select(self, event_model):
        where_clause = self._get_where_clause(event_model)

        (
            final_context_columns_clause,
            context_column_clause,
            context_id_column_clause,
            context_join_clause,
            annotated_context_columns_clause,
        ) = self._get_context_clauses(event_model)

        pgh_obj_model = event_model._meta.get_field("pgh_obj").remote_field.model
        pgh_obj_table = pgh_obj_model._meta.db_table
        update_trackers = [
            tracker for tracker in event_model.pgh_trackers if tracker.label == "update"
        ]

        curr_data_clause = "row_to_json(_event) AS _curr_data"
        prev_data_clause = """
            LAG(row_to_json(_event))
              OVER (
                PARTITION BY _event.pgh_obj_id
                ORDER BY _event.pgh_id
              ) AS _prev_data
        """
        pgh_obj_id_column_clause = "pgh_obj_id::TEXT"
        if not hasattr(event_model, "pgh_obj_id"):
            prev_data_clause = "NULL::JSONB AS _prev_data"
            pgh_obj_id_column_clause = "NULL::TEXT AS pgh_obj_id"
        elif update_trackers and update_trackers[0].row == "OLD":  # Check if Old or New
            curr_data_clause = f"""
                COALESCE(
                    LEAD(row_to_json(_event)) OVER (PARTITION BY _event.pgh_obj_id ORDER BY _event.pgh_id),
                    (SELECT row_to_json(m) FROM {pgh_obj_table} m WHERE m.id = _event.pgh_obj_id)
                ) AS _curr_data
            """
            prev_data_clause = "row_to_json(_event) AS _prev_data"

        event_table = event_model._meta.db_table
        return f"""
            SELECT
              CONCAT('{event_model._meta.label}', ':', _pgh_obj_event.pgh_id) AS pgh_slug,
              _pgh_obj_event.pgh_id,
              _pgh_obj_event.pgh_created_at,
              _pgh_obj_event.pgh_label,
              {final_context_columns_clause}
              _pgh_obj_event.pgh_obj_id,
              '{event_model._meta.label}' AS pgh_model,
              '{event_model.pgh_tracked_model._meta.label}' AS pgh_obj_model,
              (
                  SELECT JSONB_OBJECT_AGG(filtered.key, filtered.value)
                  FROM
                    (
                        SELECT key, value
                        FROM JSONB_EACH(_pgh_obj_event._curr_data::JSONB)
                    ) filtered
                  WHERE filtered.key NOT LIKE 'pgh_%%'
              ) AS pgh_data,
              (
                SELECT JSONB_OBJECT_AGG(curr.key, array[prev.value, curr.value])
                FROM
                  (
                    SELECT key, value
                    FROM JSONB_EACH(_pgh_obj_event._curr_data::JSONB)
                  ) curr
                  LEFT OUTER JOIN
                  (
                    SELECT key, value
                    FROM JSONB_EACH(_pgh_obj_event._prev_data::JSONB)
                  ) prev
                  ON curr.key = prev.key
                WHERE curr.key NOT LIKE 'pgh_%%'
                  AND curr.value != prev.value
                  AND prev IS NOT NULL
              ) AS pgh_diff,
              _pgh_obj_event.pgh_context_id,
              _pgh_obj_event.pgh_context
            FROM (
              SELECT
                pgh_id,
                pgh_created_at,
                pgh_label,
                {curr_data_clause},
                {annotated_context_columns_clause}
                {prev_data_clause},
                {context_id_column_clause},
                {context_column_clause},
                {pgh_obj_id_column_clause}
              FROM {event_table} _event
              {context_join_clause}
              {where_clause}
              ORDER BY _event.pgh_id
            ) _pgh_obj_event
        """
```

I can open a PR and update to align with project scopes as needed.


This part is not quite clear for me

```python
if not hasattr(event_model, "pgh_obj_id"):
            prev_data_clause = "NULL::JSONB AS _prev_data"
            pgh_obj_id_column_clause = "NULL::TEXT AS pgh_obj_id"
```

**Comment by bgervan** [comment]
I created a sample model with 

```python
PGHISTORY_DEFAULT_TRACKERS = (
    pghistory.InsertEvent(),
    pghistory.UpdateEvent(row=pghistory.Old),
    pghistory.DeleteEvent(row=pghistory.Old),
)
```

Insert: `name: Sample`
Update 1: `name: Sample (update1)`
Update. 2: `name: Sample (update2)`

With the above changes, the diffs are the following respectively:

```json
null

null

{
    "name": [
        "Sample",
        "Sample (update1)"
    ],
    "updated_at": [
        "2024-10-10T05:21:29.032954+00:00",
        "2024-10-10T05:21:38.175114+00:00"
    ]
}
```

So when I updated to `Sample (update2)`, I have the diff of the previous change, which is a bug with the pghistory.Old settings. For update1, I receive null, when there is a change.

With my update, the diffs are the following:

```python
In [3]: MyEvents.objects.tracks(example)[0].pgh_diff

In [4]: MyEvents.objects.tracks(example)[1].pgh_diff
Out[4]: 
{
   'name': ['Sample', 'Sample (update1)'],
   'updated_at': [
       '2024-10-10T05:21:29.032954+00:00',
       '2024-10-10T05:21:38.175114+00:00'
   ]
}

In [5]: MyEvents.objects.tracks(example)[2].pgh_diff
Out[5]: 
{
  'name': ['Sample (update1)', 'Sample (update2)'],
  'updated_at': [
       '2024-10-10T05:21:38.175114+00:00',
       '2024-10-10T05:21:43.575616+00:00'
   ]
}
```

So for insert it is still NULL, but for the updates, it shows the correct values. 

The fix is simply shifts the comparison objects from last and current, to current and next

**Comment by bgervan** [comment]
I understood what happens, that is what my fix trying to update, checking the tracker if its NEW or OLD and return the diff accordingly. 

You don't consider it as a bug when the user uses the row=pghistory.Old? I found a couple of issue with my fix, but fixed since the first implementation.

**Comment by wesleykendall** [comment]
If you are tracking `OLD` rows on update, your snapshot table lags behind the current table. The aggregate queryset that shows diffs only examines snapshot rows, so what you're seeing is expected behavior.

One alternative approach is this:

```
PGHISTORY_DEFAULT_TRACKERS = (
    pghistory.InsertEvent(),
    pghistory.UpdateEvent(row=pghistory.New),
    pghistory.DeleteEvent(row=pghistory.Old),
)
```

If you do this, the example above now looks like this:

- INSERT of col_1=a will produce tracked history of col_1=a
- UPDATE of col_1=b will track the *NEW* row *after* the change and produce a tracked history of col_1=b
- UPDATE of col_1=c will track the *NEW* row *after* the change and produce a tracked history of col_1=c

I.e. your snapshot table will actually show differences in the first two snapshot entries.

Another possibility is to filter out inserts from the snapshots table or not track them at all.

To summarize, the aggregate diff utility is just showing diffs from the snapshot table. What's been reported here is expected behavior. Altering it would introduce regressions for others and require an API break.

**Comment by wesleykendall** [comment]
Going to go ahead and close this because it seems like you are reporting expected behavior. If you have a suggestion for how we can extend it for other use cases, try opening a [discussion](https://github.com/Opus10/django-pghistory/discussions)
